counterpartylib/lib/transaction.py

```diff
@@ -487,4 +491,199 @@ def construct (db, tx_info, encoding='auto',
+                outputs_info['data'] += new_data
+                outputs_info['data_value'] += vout.nValue
+                outputs_info['data_output_count'] += 1
+                if outputs_info['data_output_type'] is not None and outputs_info['data_output_type'] != output_type:
+                    raise exceptions.TransactionError('constructed transaction contains invalid data outputs')
+                outputs_info['data_output_type'] = output_type
+        outputs_info['total_value'] += vout.nValue
+
+    return outputs_info
+
+def check_outputs(method, params, tx_hex):
+    tx_info = get_outputs_info(tx_hex)
+    #print(tx_info)
+    asked_message_type = method.split('_')[1]
+
+    if tx_info['data']:
+        
+        data_hex = binascii.hexlify(tx_info['data'])
+        data = raw_unpack_data(data_hex)
+        #print(params)
+        #print(data)
+
+        # check message type
+        returned_message_type = data.pop('message')
+        if asked_message_type != returned_message_type:
+            raise exceptions.TransactionError('invalid message type')
+
+        # asset_id to asset name (hackish)
+        asset_param_list = {
+            'asset_id': 'asset', 
+            'give_id': 'give_asset', 
+            'get_id': 'get_asset',
+            'dividend_asset_id': 'dividend_asset'
+        }
+        for asset_param in asset_param_list:
+            if asset_param in data:
+                asset_id = data.pop(asset_param)
+                param_name = asset_param_list[asset_param]
+                try:
+                    data[param_name] = util.generate_asset_name(asset_id, block_index=333501)
+                except exceptions.AssetIDError:
+                    data[param_name] = util.generate_asset_name(asset_id, block_index=333499)
+            
+        # check change address
+        if tx_info['change'][0] != None and tx_info['change'][0] != params['source']:
+            raise exceptions.TransactionError('invalid change address')
+
+        # check parameters
+        optional_params = {
+            'call_price': 0.0,
+            'description': b'',
+            'call_date': 0,
+            'callable_': False,
+            'divisible': True
+        }
+        for param in data:
+            invalid_param = param != 'message'
+            invalid_param = invalid_param and (param not in params or data[param] != params[param])
+            invalid_param = invalid_param and (param not in optional_params or data[param] != optional_params[param])
+            if invalid_param:
+                raise exceptions.TransactionError('constructed transaction contains invalid parameter: {}'.format(param))
+
+        # TODO: check all messages
+        # check destination
+        if asked_message_type == 'bet':
+            params['destination'] = params['feed_address']
+
+        # No destination
+        if asked_message_type in ['order', 'dividend', 'broadcast', 'cancel', 'destroy', 'execute', 'publish', 'rps', 'rpsresolve']:
```

Comment:
Or rather


And create new variables `parsed_destinations = tx_info['destinations'].keys()` and `desired_destination = params[destination']`, I think.